[RAM][Security Solution][Alerts] moves legacy actions migration to rulesClient

vitaliidm opened the pull request:
## Summary
- this PR is the first part of work related to conditional logic actions. The rest of PRs, will be merged after this one. As they depend on a work implemented here. [List of tickets](https://github.com/elastic/security-team/issues/2894#issuecomment-1480253677)
- addresses https://github.com/elastic/kibana/issues/151919
- moves code related to legacy actions migration from D&R to RulesClient, [details](https://github.com/elastic/kibana/issues/151919#issuecomment-1473913699)
   - similarly to D&R part, now rulesClient read APIs, would return legacy actions within rule
   - similarly, every mutation API in rulesClient, would migrate legacy actions, and remove sidecar SO
- each migrated legacy action will have also [`frequency` object](https://github.com/elastic/kibana/blob/8.7/x-pack/plugins/alerting/server/types.ts#L234-L238), that would allow to have notifyWhen/throttle on action level once https://github.com/elastic/kibana/issues/151916 is implemented, which is targeted in 8.8, right after this PR.
But before it's merged, `frequency` is getting removed in [update/bulk_edit/create APIs](https://github.com/elastic/kibana/blob/8.7/x-pack/plugins/alerting/server/rules_client/methods/update.ts#L151-L160). Hence it's not reflected in most of the tests at this point.
- cleanup of legacy actions related code in D&R
- adds unit tests for RulesClient
- keeps functional/e2e tests in D&R

Changes in behaviour, introduced in this PR:
- since, migration happens within single rulesClient API call, revision in migrated rule will increment by `1` only.
- legacy actions from sidecar SO, now will be merged with rules actions, if there any.
Before, in the previous implementation, there was inconsistency in a way how legacy and rules actions were treated.
  - On read: actions existing in rule, [would take precedence over legacy ones  ](https://github.com/elastic/kibana/blob/8.7/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_actions.ts#L94-L114)
  - On migration: SO actions [only saved](https://github.com/elastic/kibana/blob/8.7/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/rule_actions/legacy_action_migration.ts#L114). If any actions present in rule, they will be lost. Here is an example video from **main** branch
      <details>
      <summary>Here is an example video from MAIN branch, where action in rule is overwritten by legacy action</summary>
      
      https://user-images.githubusercontent.com/92328789/230397535-d3fcd644-7cf9-4970-a573-18fd8c9f2235.mov

      </details>

  So, depends on sequence of events, different actions could be saved for identical use case: rule has both legacy and existing action. 
     - if rule migrated through update API, existing in rule action will be saved
     - if rule migrated through enable/disable API, bulk edit, legacy action will be saved

  In this implementation, both existing in rule and legacy actions will be merged, to prevent loss of actions
      <details>
      <summary>Here is an example video from this PR branch, where actions are merged</summary>
      <video src="https://user-images.githubusercontent.com/92328789/230405569-f1da38e9-4e36-46a8-9654-f664e0a31063.mov" />
      </details>
      
- when duplicating rule, we don't migrate source rule anymore. It can lead to unwanted API key regeneration, with possible loss of privileges, earlier associated with the source rule. As part of UX, when duplicating any entity, users would not be expecting source entity to be changed
  
TODO:
- performance improvement issue for future https://github.com/elastic/kibana/issues/154438
- currently, in main branch, when migration is performed through rule enabling, actions not showing anymore in UI. 
Relevant ticket is https://github.com/elastic/kibana/issues/154567
I haven't fixed it in this PR, as in[ the next one ](https://github.com/elastic/kibana/pull/153113), we will display notifyWhen/throttle on action level in UI, rather than on rule level. So, once that PR is merged, actions should be displayed in new UI
  
### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

vitaliidm committed:
[RAM][Security Solution][Alerts] moves legacy actions migration to rulesClient

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

vitaliidm committed:
fix lint errors

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-to-alerting' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-to-alerting

# Conflicts:
#	x-pack/plugins/security_solution/server/lib/detection_engine/rule_actions_legacy/logic/rule_actions/legacy_rules_client_migration_hook.ts

vitaliidm committed:
runs migrations concurrently

vitaliidm committed:
fix lint errors

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
fix various issues

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-to-alerting' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
fix some of the tests

vitaliidm committed:
fix tests

vitaliidm committed:
fix tests

vitaliidm committed:
fix lint

vitaliidm committed:
fix even more failed tests

vitaliidm committed:
fix more tests

vitaliidm committed:
fix

vitaliidm committed:
add migration

vitaliidm committed:
fix failed tests

vitaliidm committed:
fix jest

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
remove migrations form DE

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-to-alerting' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
add import rules as well

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-to-alerting' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
add prebuilt rules route

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-to-alerting' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-to-alerting

vitaliidm committed:
rules bulk delete

vitaliidm committed:
clean up jest tests

vitaliidm committed:
add unit tests

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-to-alerting

# Conflicts:
#	x-pack/plugins/alerting/server/index.ts
#	x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_objects/update_prebuilt_rules.test.ts
#	x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_objects/update_prebuilt_rules.ts

vitaliidm committed:
add bulk_edit

vitaliidm committed:
migrateLegacyActions

vitaliidm committed:
add bulk_edit

vitaliidm committed:
add delete

vitaliidm committed:
remove hook from registry

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

# Conflicts:
#	x-pack/plugins/alerting/server/rules_client/lib/index.ts
#	x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts
#	x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_by_object_ids.ts

vitaliidm committed:
fix tests

vitaliidm committed:
fix e2e tests

vitaliidm committed:
fix types

vitaliidm committed:
add enable/disable, refactoring

vitaliidm committed:
remove leftover code

vitaliidm committed:
fix jest tests

vitaliidm committed:
fix duplicvate test

vitaliidm committed:
fix jests tests

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
please, work

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v2' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v2

vitaliidm committed:
fix uuid

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
refactoring

vitaliidm committed:
find

vitaliidm committed:
cleanup hook code

vitaliidm committed:
Merge branch 'elastic:main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
fix tests for detections

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v2' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v2

vitaliidm committed:
add migration to single update method

vitaliidm committed:
update more methods

vitaliidm committed:
update the rest of methods

vitaliidm committed:
Update one_action.json

vitaliidm committed:
add uuid

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
add some tests

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v2' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v2

vitaliidm committed:
add tests for single methods

vitaliidm committed:
add the rest of the tests for edit

vitaliidm committed:
refactoring

vitaliidm committed:
add more tests for read methods

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
test for fonmd

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v2' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v2

vitaliidm committed:
D&R refactoring + tests

vitaliidm committed:
rf leftovers

vitaliidm committed:
return file

vitaliidm committed:
more teeests

vitaliidm committed:
remove unnecessary TODO

vitaliidm committed:
tests fixes

vitaliidm committed:
fix failed tests

vitaliidm committed:
remove TODO

vitaliidm committed:
new tests format

vitaliidm committed:
initial commit

vitaliidm committed:
remove onluyy in tests

vitaliidm committed:
refacotring

vitaliidm committed:
fixes

vitaliidm committed:
fixes leagcy actions references

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
additional tests

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v2' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v2

spong commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_actions.ts:
```diff
@@ -63,25 +59,16 @@ export const transformToAlertThrottle = (throttle: string | null | undefined): s
  * @param legacyRuleActions Legacy "side car" rule actions that if it detects it being passed it in will transform using it.
```
Leftover JSDoc param:

```suggestion
```

spong reviewed (approved):
Code reviewed detection-rule area files and changes LGTM! 👍 

Thank you for the detailed description and outline of the plan here @vitaliidm -- appreciate it! Note: I didn't extensively test these changes locally, just some smoke testing, but our area code changes looked good, and the remaining failing tests appear to be flake, so no concern there.

FYI: Those flakey Related Integration test failures will be skipped as of https://github.com/elastic/kibana/pull/154662

XavierM commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/format_legacy_actions.ts:
```diff
@@ -0,0 +1,137 @@
+  const actionSavedObjects = results.flat().flatMap((r) => r.saved_objects);
+
+  if (errors.length) {
+    throw new AggregateError(errors, 'Error fetching rule actions');
```
why are we using `AggregateError`?

vitaliidm replied:
this is part of the original Security SOlution code, that migrated actions from https://github.com/elastic/kibana/blob/4072cf2d3d83995881ced99c4ef900a853813b33/x-pack/plugins/security_solution/server/lib/detection_engine/rule_actions_legacy/logic/rule_actions/legacy_get_bulk_rule_actions_saved_object.ts#L60-L62

I think, because there are multiple errors, `AggregateError` is used for its purpose

XavierM commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/transform_legacy_actions.ts:
```diff
@@ -0,0 +1,59 @@
+/*
```
I know that you have test on the `format_legacy_actions.test.ts` we think  that 's nice to have it per file and function make it easier to know what we can change if we need to.

XavierM commented on x-pack/plugins/alerting/server/rules_client/methods/create.ts:
```diff
@@ -103,7 +103,7 @@ export async function create<Params extends RuleTypeParams = never>(
   // TODO https://github.com/elastic/kibana/issues/148414
   // If any action-level frequencies get pushed into a SIEM rule, strip their frequencies
-  const firstFrequency = data.actions[0]?.frequency;
+  const firstFrequency = data.actions.find((action) => action?.frequency)?.frequency;
```
Tell if I am wrong here, but @e40pud should be able to delete that completely. I am wondering if it will be better to avoid the change for now.???

vitaliidm replied:
Yes, after Zhenya's changes this should be removed.

I've made change here to avoid broken tests on Security Solution side.
Since the change of how we are merging existing actions and legacy ones:

> - legacy actions from sidecar SO, now will be merged with rules actions, if there any.
Before, in the previous implementation, there was inconsistency in a way how legacy and rules actions were treated.
  - On read: actions existing in rule, [would take precedence over legacy ones  ](https://github.com/elastic/kibana/blob/8.7/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_actions.ts#L94-L114)
  - On migration: SO actions [only saved](https://github.com/elastic/kibana/blob/8.7/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/rule_actions/legacy_action_migration.ts#L114). If any actions present in rule, they will be lost.


Existing actions do not have frequencies, but migrated already have(introduced in this PR). So, I just made a small amend how frequency is searched in a list of actions. Since, after merging existing and legacy, existing action are first in order, but don't have frequencies to be removed

XavierM reviewed (commented):
I think we can create a general helpers function to do that and can return `hasLegacyActions` and `resultedActions`. Then we can use it on every methods. what do you think?
```ts
let resultedActions: RawRule['actions'] = [];
    let resultedReferences: SavedObjectReference[] = [];
    let hasLegacyActions = false;

    // migrate legacy actions only for SIEM rules
    if (attributes.consumer === AlertConsumers.SIEM) {
      const migratedActions = await migrateLegacyActions(context, {
        ruleId: id,
        actions: attributes.actions,
        references,
        attributes,
      });

      resultedActions = migratedActions.actions;
      resultedReferences = migratedActions.references;
      hasLegacyActions = migratedActions.hasLegacyActions;
    }
    ```

XavierM reviewed (commented):
@vitaliidm, I reviewed the PR and the code looks good but I did not play with it or test it yet. I think it will be useful to run some queries on 7.last and use these SOs in this PR and  local environment to make use everything work as expected.

vitaliidm committed:
Update x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_actions.ts

Co-authored-by: Garrett Spong <spong@users.noreply.github.com>

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v2' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v2

vitaliidm committed:
fix broken tests

vitaliidm committed:
add additional tests

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

# Conflicts:
#	x-pack/plugins/alerting/server/rules_client/methods/bulk_enable.ts
#	x-pack/plugins/alerting/server/rules_client/methods/update.ts
#	x-pack/plugins/alerting/server/rules_client/methods/update_api_key.ts
#	x-pack/plugins/alerting/server/rules_client/tests/bulk_delete.test.ts
#	x-pack/plugins/alerting/server/rules_client/tests/update.test.ts

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

XavierM commented:
> I think we can create a general helpers function to do that and can return `hasLegacyActions` and `resultedActions`. Then we can use it on every methods. what do you think?
> 
> ```ts
> let resultedActions: RawRule['actions'] = [];
>     let resultedReferences: SavedObjectReference[] = [];
>     let hasLegacyActions = false;
> 
>     // migrate legacy actions only for SIEM rules
>     if (attributes.consumer === AlertConsumers.SIEM) {
>       const migratedActions = await migrateLegacyActions(context, {
>         ruleId: id,
>         actions: attributes.actions,
>         references,
>         attributes,
>       });
> 
>       resultedActions = migratedActions.actions;
>       resultedReferences = migratedActions.references;
>       hasLegacyActions = migratedActions.hasLegacyActions;
>     }
>     ```
> ```

@vitaliidm and I talked about it. I think I miss understand him and he did it like this at the firs implementation and I ask him to have test to make sure that every method is calling the migration. So we decide to move with this pattern for now and we will create an issue to refactor the way @vitaliidm implemented at first. Sorry for the confusion, I should have known that you were right at the first place!!!

banderror commented:
I just read the description and all the associated tickets, and it all sounds just awesome @vitaliidm. Appreciate your attention to detail and edge cases 🙏 

I was going to review this one, and Xavier asked me today to take a look too, so I'll do a proper pass tomorrow morning. Super excited!

vitaliidm committed:
add frequency to existing actiopns

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

yctercero commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/format_legacy_actions.ts:
```diff
@@ -0,0 +1,137 @@
+/**
+ * @deprecated Once we are confident all rules relying on side-car actions SO's have been migrated to SO references we should remove this function
+ */
+export const legacyGetBulkRuleActionsSavedObject = async ({
```
Note: one thing I wish we'd done the first time around writing the migration code is to have better documented _everything_ . So even though it may seem obvious right now what each of these do as we're working on it, you (general "you") will soon forget. 

Maybe adding a description of what each function does, however obvious it may seem, will help our future selves. Some additional documenting that may help is a note as to why we only search by `siem-detection-engine-rule-actions` and not also `siem.notifications`. I outlined it in this PR - pre 7.16, even rules created with no actions or actions for every rule run, generate a `siem-detection-engine-rule-actions` but not a `siem.notifications` SO, so we cannot rely on `siem.notifications` alone.

vitaliidm replied:
thanks for the suggestion.
I put some additional comments to functions and reference to that PR as well

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
fix tests

banderror commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/delete_rules.ts:
```diff
@@ -216,6 +225,10 @@ export default ({ getService }: FtrProviderContext): void => {
 
         // Expect that we have exactly 0 legacy rules after the deletion
         expect(bodyAfterDelete.total).to.eql(0);
+
+        // legacy sidecar action should be gone
+        const sidecarActionsPostResults = await getLegacyActionSO(es);
+        expect(sidecarActionsPostResults.hits.hits.length).to.eql(0);
```
👍

banderror reviewed (commented):
Reviewed the diff but didn't test the app manually -- I'm not so familiar with the legacy actions data model, so I'd rely on existing integration tests for legacy actions and rule management API endpoints + review from the Platform side. Thank you for extending and fixing the tests @vitaliidm.

I left a bunch of comments, mostly minor. My main question is this: https://github.com/elastic/kibana/pull/153101#discussion_r1165359219

vitaliidm committed:
fix seime tests

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v4' of ssh://github.com/vitaliidm/kibana into alerts/legacy-actions-migration-v4

vitaliidm committed:
update "update" method

vitaliidm committed:
fix bulk_edit

vitaliidm committed:
Merge branch 'alerts/legacy-actions-migration-v4' into alerts/legacy-actions-migration-v2

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/__mocks__/request_responses.ts:
these mocks are still used for legacy-notifications API

vitaliidm committed:
CR feedback

vitaliidm committed:
CR

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/retrieve_migrated_legacy_actions.mock.ts:
```diff
@@ -0,0 +1,357 @@
+/**
+ * @deprecated Once we are confident all rules relying on side-car actions SO's have been migrated to SO references we should remove this function
+ */
+export const legacyGetHourlyNotificationResult = (id = '456', ruleId = '123') => ({
```
added typings

vitaliidm commented:
@XavierM 

So,  I did the following updates as agreed since adv. correlation meeting and recent discussions:
- `throttle` and `notifyWhen` will be set to `undefined` during legacy migration for all methods apart from `bulkEdit`.
In current state bulkEdit still receives operations from SecuritySolution for setting `throttle` and `notifyWhen` on the rule level during a single call. It breaks tests. I guessing it should be fine though. As the next @e40pud  PR would amend this method and frequencies will be set per action level, instead of separate bulk edit operations.
- I moved code, mentioned in https://github.com/elastic/kibana/pull/153101#pullrequestreview-1378124040 to `migrateLegacyActions`, so now it's encapsulated there. I also changed necessary tests for this
- I haven't yet touched `muteAll` on read, but I've got [a separate PR](https://github.com/elastic/kibana/pull/154860/commits/20cd0b883de3db740a8e5c98be3721e3c6c7a5c5), that runs[ a build for it](https://github.com/elastic/kibana/pull/154860/commits/20cd0b883de3db740a8e5c98be3721e3c6c7a5c5). If it passes, let's include this change in this PR. And we might need to do additional testing on that once Zhenya's PR is merged and Snooze introduced in Security Solution. Or maybe we could do it after these PRs, to avoid rare(or even impossible in real life) situations we discussed on call(muted rule with SO actions)

vitaliidm committed:
add tests for enabling rule

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
revert update key API

vitaliidm committed:
revert tests

vitaliidm committed:
revert unnecesary methods

vitaliidm committed:
fix enable

vitaliidm commented:
- removed migration for a number of methods and left only for the following: single and bulk update, enable/disable, delete.
- in enable method, we calling create with overwrite=true to prevent AAD issues with key encrypting
- added test that covers successful rule execution after its enabling

vitaliidm committed:
cleanup

vitaliidm committed:
update enable test

vitaliidm committed:
enable method

vitaliidm committed:
Revert "enable method"

This reverts commit 115857f08027658b1c757b1a6c1045eab400f628.

vitaliidm committed:
fix api key regeneration

XavierM commented on x-pack/plugins/alerting/server/rules_client/methods/disable.ts:
```diff
@@ -4,13 +4,14 @@
  * 2.0; you may not use this file except in compliance with the Elastic License
  * 2.0.
  */
+import type { SavedObjectReference } from '@kbn/core/server';
```
We decided to not touch this endpoint.

XavierM commented on x-pack/plugins/alerting/server/rules_client/methods/bulk_disable.ts:
```diff
@@ -4,7 +4,6 @@
  * 2.0.
  */
-
 import { KueryNode, nodeBuilder } from '@kbn/es-query';
```
we decided to not touch this endpoint

XavierM commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/migrate_legacy_actions.ts:
```diff
@@ -0,0 +1,98 @@
+ * @param params
+ * @returns
+ */
+export const migrateLegacyActions: MigrateLegacyActions = async (
```
Can we add a try/catch here and log the error to not block the main method? what do you think?

vitaliidm replied:
added

XavierM commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/transform_legacy_actions.ts:
```diff
@@ -0,0 +1,59 @@
+    {}
+  );
+
+  if (isEmpty(actionReference)) {
```
do we need to throw an error, maybe we can just ignore it and return empty array?

vitaliidm replied:
added try/catch in `formatLegacyActions` and `migrateLegacyActions` on top level. So any thrown errors will be caught and logged

XavierM commented on x-pack/plugins/alerting/server/rules_client/lib/siem_legacy_actions/format_legacy_actions.ts:
```diff
@@ -0,0 +1,145 @@
+  const actionSavedObjects = results.flat().flatMap((r) => r.saved_objects);
+
+  if (errors.length) {
+    throw new AggregateError(errors, 'Error fetching rule actions');
```
let's mute the error, we should log it for sure!

vitaliidm replied:
added try/catch in `formatLegacyActions` and `migrateLegacyActions` on top level. So any thrown errors will be caught and logged

XavierM reviewed (approved):
Please make sure, to mute any errors during migration/formatting, so we do not block our user form the main method that they wanted to do.

Also I tested it locally and running and everything is working as expected and I did not see any re-generation of API key.

vitaliidm committed:
Update route.ts

vitaliidm committed:
Update one_action.json

vitaliidm committed:
Merge branch 'main' into alerts/legacy-actions-migration-v2

vitaliidm committed:
mute error for RAM migrations

vitaliidm committed:
add test for muting erorrs

yctercero reviewed (approved):
Codeowners for security-solution-platform LGTM. I haven't gotten to test all the cases on my local yet, but do not want to block this as I know other PR's are waiting on this part to go in. 

I saw we do increment the rule version - is that also true for migration on a read? Would that appear strange to the user that on a read, their rule appears to have been modified? (I know technically it was, but they won't necessarily know why or in what way).

vitaliidm committed:
additonal tests for muting errors

vitaliidm commented:
> I saw we do increment the rule version - is that also true for migration on a read? Would that appear strange to the user that on a read, their rule appears to have been modified? (I know technically it was, but they won't necessarily know why or in what way).

@yctercero 
Rule revision is incremented on every update, independent to migration of legacy rules. And not incremented on read at all.

> since, migration happens within single rulesClient API call, revision in migrated rule will increment by 1 only.

In current implementation in Security Solution, when user updates rule, we do 2 updates under the hood:
1. migrate legacy actions, if any present
2. actual rule update

So, in legacy actions migration case revision is getting incremented by 2. Here is a PR, when revision increment was introduced: https://github.com/elastic/kibana/pull/147398

But since, migration is moved to rulesClient, we perform only single update. So, revision will be incremented by `1` only.
